Editor: Operate on template CPT posts and add a default template with post title and content blocks.

epiqueras opened the pull request:
cc @youknowriad 

Closes #16281

## Description

This PR builds on @aduth 's work on #16402 to add backwards compatible support for making the editor operate on and persist to template CPT posts instead of the actual edited post.

It adds a single, default, editable template to the "post" post type which contains a new post-title block, which persists to the edited post's title, followed by a new post-content block, which persists to the edited post's `post_content`. Note that this post-content block produces no markup on `save` so that the template only stores the location of `post-content` relative to the other blocks in the template. When the template is rendered by a theme, either explicitly or dynamically like in #4659, a render callback for the block could fill in the viewed post's content.

Future work will be needed to allow users or theme authors to create a template hierarchy and apply them to specific posts.

## How has this been tested?

Posts were edited and saved and it was verified that the template updated correctly when making changes outside of the post-content block, as well as the title for the post-title block, and the content for the post-content block.

## Screenshots

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/19157096/60903807-9f3d3380-a237-11e9-84fb-bc627a09dac7.gif)

## Types of Changes

*New Feature:* Operate on template CPT posts and add a default template with post title and content blocks.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [ ] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [ ] I've included developer documentation if appropriate.

epiqueras committed:
Block Library: Add new post-title block.

epiqueras committed:
Edit Post: Set a default post template that uses the core/post-title block.

Remove the `PostTitle` component from the `VisualEditor` component.

epiqueras committed:
Block Library: Add new post-content block and use it in the default post template.

epiqueras committed:
Templates: Implement persistence using a template CPT.

epiqueras committed:
Templates: Integrate with demo content.

epiqueras committed:
Editor: Integrate post-title and post-content with custom sources.

epiqueras committed:
Editor: Keep compatibility with current template format.

epiqueras committed:
Editor: If a template is supplied, apply it if it's editable or if it's static and the post is new.

epiqueras committed:
Post Title Block: Sync value with post slug.

epiqueras commented:
I've rebased the zoom prototyping out of this PR and into #16578 .

youknowriad commented on packages/editor/src/store/actions.js:
```diff
@@ -474,11 +478,24 @@ export function* savePost( options = {} ) {
 		'getEditedPostContent'
 	);
 
+	const template = ( yield select( STORE_KEY, 'getEditorSettings' ) ).template;
```
Oh, so you're using the current `template` editor setting as the actual "FSE template". The `template` property is meant to be the `post_content` template while the actual FSE template is something that get stored separately. There's a question to be asked about "naming confusion" and also whether we may want to absorb one use-case in another in the future but I think for now we may want to consider them as two separate entities.

Basically what I had in mind is something like that:

 - a separate `template` prop in the  `EditorProvider` for the FSE template
 - The same way we have `currentPost` and `edits` for the post in the reducer, I think we'd need `currentTemplate` and `templateEdits` maybe.
 - a boolean prop or something like that in the `EditorProvider` to switch between FSE mode (showing the template blocks) and current mode (showing a PostTitle input and the PostContent blocks as root level)
 - Both `core/post-content`  and `core/post-title` blocks should be dynamic blocks that render the actual post title and post content in the frontend.
 - Basically `post-content.php` file shouldn't have to be touched to be loaded in an editor that has a template as well.

Does that make sense?

epiqueras replied:
Yeah, that was what I was planning, except I wasn't sure if I should change the way the template was being injected and was going to leave that for last so we could discuss it further.

From my zoom prototyping:

>I started by trying to support insertions, replacements, and removals, but I quickly realized that it would take a pretty widespread refactor due to a lot of assumptions we make in the logic for getting insertion indexes and other data. I don't think the need for this feature warrants those changes. This makes me think that we might have to try a different approach for #16485 where the editor just switches between editing the template and editing post content with a static post title component like it does now.

So for next steps here I see:

- [ ] Give the template its own reducer and make all the necessary refactors. (Use the `'block_editor_settings'` filter to send it to the client in its own, new key.)
- [ ] Implement the "zoom" as a toggle between setting up the editor with the template or post.
- [ ] Implement the render callbacks of the post title and post content blocks.

epiqueras committed:
Editor: Make post templates a separate feature from block templates and add a toggle between template and post mode.

epiqueras commented:
- [x] Give the template its own reducer and make all the necessary refactors. (Use the `'block_editor_settings'` filter to send it to the client in its own, new key.)
- [x] Implement the "zoom" as a toggle between setting up the editor with the template or post.
- [ ] Implement the render callbacks of the post title and post content blocks.

>https://github.com/WordPress/gutenberg/pull/16565/commits/b09a54d8cf95c0bd91e24ad430b9373589ee4c96 Editor: Make post templates a separate feature from block templates and add a toggle between template and post mode.

epiqueras committed:
Templates: Add render callbacks for post title and content blocks and override the single post template with its editable counterpart.

epiqueras commented:
- [x] Give the template its own reducer and make all the necessary refactors. (Use the `'block_editor_settings'` filter to send it to the client in its own, new key.)
- [x] Implement the "zoom" as a toggle between setting up the editor with the template or post.
- [x] Implement the render callbacks of the post title and post content blocks.

>https://github.com/WordPress/gutenberg/pull/16565/commits/6136f29f2fc4ef74b6b8511de6a2e6b51ce5a603 Templates: Add render callbacks for post title and content blocks and override the single post template with its editable counterpart.

youknowriad committed:
Adding a post date block

epiqueras committed:
Editor: Fix template post setup.

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -167,11 +169,24 @@ export function* setupEditor( post, edits, template ) {
 	}
 
-	let blocks = parse( content );
+	let blocks = ( edits && edits.blocks ) || parse( content );
```
`edits` is the accumulation of post property values we plan to send as updates to the server.

With your revision here, can you define what purpose `edits` serves?

jorgefilipecosta committed:
Add: Post content does not show theme blocks (#16676)

aduth commented on packages/editor/src/store/actions.js:
```diff
@@ -167,11 +169,24 @@ export function* setupEditor( post, edits, template ) {
+			( block ) => block.name === 'core/post-content'
+		);
+		if ( postContentBlock ) {
+			postContentBlock.innerBlocks = isNewPost ? // Apply block (post content) template.
```
The assignment of `innerBlocks` is mutating state.

mcsf committed:
Add Site Title block and support for site options attribute source

mtias committed:
Fix date php code.

jorgefilipecosta committed:
Enable navigation menu in the inserter. Make menu block part of the theme block category.

jorgefilipecosta committed:
Update initial template to a more advanced design that includes navigation and post date

gziolo committed:
Full Site Editing: Add view editing modes picker (#16680)

* Add view editing modes picker

* Move view editing mode to preferences in state

* Hide label from displaying in the header for view mode picker

* Add preview mode

* Fix some inconsistencies

jorgefilipecosta committed:
Try template part

epiqueras committed:
Editor: Change save copy for new editing modes.

mtias committed:
Move view mode to the center.

youknowriad committed:
Remove nested appenders

mtias committed:
Add all modes.

mcsf committed:
Site Title: Add color support

- Rough pass for demo purposes, drawing from the example of core/heading.

- Switches from PlainText to RichText, even though any inline formats in
  the content will be escaped by the server, leading to titles with
  visible HTML tags.

youknowriad committed:
Try adding the new post block

mtias committed:
Add icons to new blocks.

epiqueras committed:
Templates: Show admin UI.

jorgefilipecosta committed:
Add template part saving

mtias committed:
Add icons for new blocks.

youknowriad committed:
Refactor mode editing

mtias committed:
COVER: Allow all blocks in inner blocks.

epiqueras committed:
Editor: Support nested post content blocks.

gziolo committed:
Disable template blocks in preview mode

jorgefilipecosta committed:
Add template part name

gziolo committed:
Disable movers for post content block as well when in preview mode

gziolo committed:
Add new focus mode for the editor

mtias committed:
Update site title font size.

mcsf committed:
Design mode: render as faded-out boxes

jorgefilipecosta committed:
Add logic to update template part ids saved on the template during the first template part save

epiqueras committed:
Editor: Add template part view editing modes.

jorgefilipecosta committed:
Use saveEntityRecord to save template parts

epiqueras committed:
Better template part mode.

epiqueras committed:
Fix template part saving.

epiqueras committed:
Fix show post title.

jorgefilipecosta committed:
Fixed templatePart hasInnerBlocks definition

mtias committed:
Add emoji to modes.

mtias committed:
More design mode CSS.

mcsf committed:
Design mode: add mock columns gradient

mtias committed:
Constrain frame editor of template parts.

jorgefilipecosta committed:
Add template part preloading mechanism

mcsf committed:
Site Title: Ascribe `theme` category

ellatrix committed:
Template Locking: readonly / disable editing attributes (without rewrite)

jorgefilipecosta committed:
Don't select template part on load

gziolo committed:
Extend template mode lock to title and date blocks

ellatrix commented:
Merged part of #16678 as 19adb5ca09fbb109a98b2ed7502fd8d008448cd4. This does not include a small rewrite of `templateLock`, to reduce the amount of file changes on this branch.

youknowriad committed:
Avoid race conditions when running multiple animations in a row

ellatrix committed:
Enable readonly post content

ellatrix commented:
Read-only mode enabled on post content in edd9eeb.

gziolo committed:
Add locking for post blocks when in template mode

ellatrix committed:
Revert "Enable readonly post content"

This reverts commit edd9eeb8d484c2f63bd5f6fadfae235468906cd2.

youknowriad committed:
Fix animation retriggered on each change

mcsf committed:
Design mode: draw grid lines (hardcoded for gutenberg-theme)

epiqueras committed:
Cache edits across modes.

ellatrix committed:
Implement template locking

epiqueras committed:
Fix demo template.

mcsf committed:
Design mode: correct and widen grid lines

mcsf committed:
Design mode: invert colours; correct width

mcsf committed:
Design mode: Background colour for lateral spaces

ellatrix committed:
Fix a bunch of locking stuff

jorgefilipecosta committed:
Add template part display name logic

ellatrix committed:
Locking: fix problem with movers

epiqueras commented:
Now Post Content:

```jsx
/**
 * WordPress dependencies
 */
import { EntityHandlers } from '@wordpress/editor';

export default function PostContentEdit() {
	return (
		<EntityHandlers handles={ { content: true, blocks: true } } />
	);
}

```

epiqueras commented:
Closing this as it was just a prototype and most of the functionality has already been merged in more granular PRs. See:

https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Full%20Site%20Editing
